org.eclipse.xtext.java/src/org/eclipse/xtext/java/resource/JavaDerivedStateComputer.xtend

```diff
@@ -124,14 +131,34 @@ class JavaDerivedStateComputer {
 		val data = resourceDescriptionsProvider.getResourceDescriptions(resource.resourceSet)
 		if (data === null)
 			throw new IllegalStateException("no index installed")
-		// TODO use container manager
-		val nameEnv = new IndexAwareNameEnvironment(resource, classLoader, data, stubGenerator, classFileCache)
-		val compiler = new Compiler(nameEnv, DefaultErrorHandlingPolicies.proceedWithAllProblems(), resource.compilerOptions, [
-			for (cls : it.classFiles) {
-				val key = QualifiedName.create(CharOperation.toStrings(cls.compoundName))
-				classFileCache.computeIfAbsent(key, [name|new ClassFileReader(cls.bytes, cls.fileName)])
+		
+		val (List<String>, Map<String, byte[]>)=>void initializer = [ topLevelTypes, classMap |
+			val inMemClassLoader = new InMemoryClassLoader(classMap, classLoader)
+			for (topLevel : topLevelTypes) {
+				try {
+					val builder = new JvmDeclaredTypeBuilder(new BinaryClass(topLevel, inMemClassLoader),
+						new ClassFileBytesAccess(), inMemClassLoader)
+					val type = builder.buildType
+					resource.contents += type
+				} catch (Throwable t) {
+					throw new IllegalStateException("could not load type '" + topLevel + "'", t)
+				}
 			}
-			if (Arrays.equals(it.fileName, compilationUnit.fileName)) {
+		] 	
+		val wasCached = classFileCache.popCompileResult(compilationUnit.fileName, initializer)
+		
+		if (!wasCached) {
+			// TODO use container manager
```

Comment:
could you elaborate: use it to do what?
I assume that this might ot be enough information for someone else who might want to resolve this

Reply:
The comment was there already. It's about creating a classloader that respects the container semantics.

Reply:
Done